### ISSUE TITLE
`cargo add` refuses to add new non-vendored packages due to source replacement

### ISSUE THREAD

**Issue by k3d3** [rust-lang/cargo#10729]
### Problem

When a vendored directory is established, `cargo add` no longer adds new packages. Instead, it tries to translate a package name into a package that already exists in the vendored directory.

In this case, I've added `xml-rs`, then proceeded to vendor my dependencies. Now, no matter what I try to `cargo add`, it always tries to add `xml-rs`.

I've even tried to run `cargo +beta add cbindgen --registry crates-io, and it still gives me the same behaviour. The only way I can get `cargo add` to work is to remove `.cargo/config` and place it back afterwards.

![image](https://user-images.githubusercontent.com/1247908/171969035-42f5fbbe-bd31-42a9-8487-a173331d09bb.png)

### Steps

```sh
cargo new example
cd example
cargo +beta add xml-rs  # The same behaviour is exhibited in +nightly as well
cargo vendor ./vendor

# (add the output to .cargo/config.toml)

cargo +beta add --build cbindgen  # Doesn't need to be a build dependency, and "cbindgen" can be anything, even crates that don't exist
cargo +beta add cbindgen --registry crates-io  # This also does not work
```

### Possible Solution(s)

I assume the reason for this is because what's added in `.cargo/config.toml` is source replacement, meaning any dependencies that try to access `crates-io` will access the vendored directory instead. I think I understand *why* `cargo add` does what it does here, however I feel it's fairly unintuitive combined with vendoring.

I haven't figured out any way to work around this (at least, in a way that I can continue to use `cargo add`). I've tried to add to config.toml:

```toml
[source.actual-crates]
registry = "https://github.com/rust-lang/crates.io-index"
```
(I've also tried git in place of https, and adding .git to the end of the URL)

and if I try to use `cargo add cbindgen --registry actual-crates`, I get:

```
error: source `actual-crates` defines source registry `crates-io`, but that source is already defined by `crates-io`
note: Sources are not allowed to be defined multiple times.
```

If I put the following into cargo.toml:
```toml
[registries]
actual-crates = { index = "https://github.com/rust-lang/crates.io-index" }
```

and run the same `cargo add` command, I get the same behaviour:
```
warning: translating `cbindgen` to `xml-rs`
      Adding xml-rs v0.8.4 to dependencies.
```

For now, I can easily add my dependencies by modifying `Cargo.toml` myself, however the `cargo add` command is very convenient, especially when introducing Rust to a team that's new to the language.

### Notes

While I am identifying this as a bug (or at least, a problem), I'm not sure what the best way is to fix it.

Making `cargo add` always grab from crates.io, despite a configured source replacement, feels like the wrong way to go about it because it has the potential to break other setups, especially ones that use mirrored (or alternative) registries.

Perhaps a flag could be given to `cargo add`? Maybe `cargo add --no-source-replacement` or something less wordy?

Alternatively, maybe there should be a new command `cargo vendor-add` that calls cargo-add code, but without the source replacement? 

### Version

```text
I've tested this on both beta and nightly cargo.

Beta's output:
cargo 1.62.0-beta.3 (4751950cc 2022-05-27)
release: 1.62.0-beta.3
commit-hash: 4751950ccc948c07047e62c20adf423d7e5f668c
commit-date: 2022-05-27
host: x86_64-pc-windows-msvc
libgit2: 1.4.2 (sys:0.14.2 vendored)
libcurl: 7.80.0-DEV (sys:0.4.51+curl-7.80.0 vendored ssl:Schannel)
os: Windows 10.0.19044 (Windows 10 Pro) [64-bit]

Nightly's output:
cargo 1.63.0-nightly (38472bc19 2022-05-31)
release: 1.63.0-nightly
commit-hash: 38472bc19f2f76e245eba54a6e97ee6821b3c1db
commit-date: 2022-05-31
host: x86_64-pc-windows-msvc
libgit2: 1.4.2 (sys:0.14.2 vendored)
libcurl: 7.83.1-DEV (sys:0.4.55+curl-7.83.1 vendored ssl:Schannel)
os: Windows 10.0.19044 (Windows 10 Pro) [64-bit]
```

**Comment by Muscraft** [comment]
It looks like this happens because a fuzzy search for a `PathSoruce` returns all packages and `cargo_add` enables fuzzy searching in `get_latest_dependency`. 

The simple solution would be to turn off fuzzy searching but I don't know if that is the correct course of action as it would stop `linked_hash_map` from being turned into `linked-hash-map`.

**Comment by k3d3** [comment]
Hmm, yeah that would at least get rid of the warning message, though you're right, that would cause other issues.

Though even if that were changed, wouldn't that cause `cargo add` to outright fail, rather than adding the crate dependency to Cargo.toml?

**Comment by Muscraft** [comment]
Yeah it would cause it to outright fail. I think throwing an error here is the best course of action as bypassing a source replace feels wrong as it wouldn't be vendored then. You would have to run `cargo add` then `cargo vendor ./vendor`. 

I personally think throwing an error if their is a source replace may be the best course of action and then have a community subcommand to deal with vendor. This is just my opinion and one of the maintainers will probably have a better idea for the path forward.

**Comment by epage** [comment]
I'm a bit annoyed with the disparate implementations of fuzzy searching.  Its difficult to provide a consistent user experience and is easy to make mistakes like this where you assume "its a registry, of course fuzzy searching is safe" when its easy to overlook the interaction with other features like source replacements.

Fundamentally, I think that is what needs to be fixed though to fix that would probably require a major refactoring to keep the current, expected behavior working.  This would also make solving #10680 less ambiguous

As for vendoring in general, we had not yet fleshed out that use case (see #10472).  Mind creating a separate issue for this?  I had missed that when creating the initial batch of cargo-add issues.

**Comment by k3d3** [comment]
Sure! So just to make sure I'm on the same page, this issue should be related to the fuzzy searching, and a new issue should be made for the cargo-add-with-vendoring use case?

**Comment by epage** [comment]
> So just to make sure I'm on the same page, this issue should be related to the fuzzy searching, and a new issue should be made for the cargo-add-with-vendoring use case?

Yes, let's keep this issue about the fuzzy searching

**Comment by epage** [comment]
I'd propose we do the following
1. Combine the query with fuzzy query methods, where separate
2. Switch from a bool to an enum (`Strict`, `Fuzzy`)
3. Add the following cases to the enum: `Normalize` and `Suggestions`
4. Explore replacing `Fuzzy` with a mixture of other enum cases

**Comment by JosiahParry** [comment]
I am presently running into this issue and I have no idea how to find my way out.

**Comment by LuuuXXX** [comment]
claim first, question it later~ @rustbot claim

**Comment by weihanglo** [comment]
I didn't quite follow. Does #13281 resolve this?

**Comment by LuuuXXX** [comment]
Yes, it’s solved. You can visually check the changes from the tests：https://github.com/LuuuXXX/cargo/tree/640c0774dae2548cf9e8c5f4fc38790d421d5783/tests/testsuite/cargo_add/add_no_vendored_package_with_vendor
https://github.com/LuuuXXX/cargo/tree/640c0774dae2548cf9e8c5f4fc38790d421d5783/tests/testsuite/cargo_add/add_no_vendored_package_with_alter_registry

**Comment by weihanglo** [comment]
Thanks for the reply. <https://github.com/rust-lang/cargo/issues/10729#issuecomment-1147444800> seems to scope this issue within fuzzy search. Close as resolved via #13281.

**Comment by quagmt** [comment]
Hi, is this error resolved? I'm encountering the same issue when using vendor and `cargo add`. I need to add the dependency manually to `Cargo.toml` and run `cargo vendor` instead.

```
❯ cargo add tracing-subscriber --features json 
error: the crate `tracing-subscriber` could not be found in registry index.
```
Here's my config
```
[source.crates-io]
replace-with = "vendored-sources"

[source.vendored-sources]
directory = "vendor"
```

cargo version
```
❯ cargo version --verbose                                                                           
cargo 1.83.0 (5ffbef321 2024-10-29)
release: 1.83.0
commit-hash: 5ffbef3211a8c378857905775a15c5b32a174d3b
commit-date: 2024-10-29
host: x86_64-unknown-linux-gnu
libgit2: 1.8.1 (sys:0.19.0 vendored)
libcurl: 8.9.0-DEV (sys:0.4.74+curl-8.9.0 vendored ssl:OpenSSL/1.1.1w)
ssl: OpenSSL 1.1.1w  11 Sep 2023
os: Ubuntu 24.4.0 (noble) [64-bit]

**Comment by weihanglo** [comment]
> > So just to make sure I'm on the same page, this issue should be related to the fuzzy searching, and a new issue should be made for the cargo-add-with-vendoring use case?
> 
> Yes, let's keep this issue about the fuzzy searching

@quagmt see <https://github.com/rust-lang/cargo/issues/10729#issuecomment-1147444800> and <https://github.com/rust-lang/cargo/issues/10729#issuecomment-1974850901>.
#10926 seems quite relevant but not immediately clear whether it is connected to `cargo vendor`.

**Comment by quagmt** [comment]
@weihanglo I see. Sadly [the issue](https://github.com/rust-lang/cargo/issues/10926) was opened 2 years ago and has no activities. I think there might be some connections as it also doesn't use the `crates.io` registry anymore when using `vendor`

**Comment by k3d3** [comment]
I believe a new issue should be made, since IMO the problem in the title doesn't match the issues others have mentioned, and (as far as I can tell) has not been solved.

The problem, at least as I define it, is that there is no functionality to add new packages into a vendored directory if you use source replacement, e.g. if you're using a vendored directory and want to add cbindgen, a package not already in the vendored directory, you can't. `cargo add` being restricted to only packages you've already added doesn't make much sense to me from a UX perspective (though I understand the underlying reasons why it works the way it does).

However, it was decided that this issue should be about the fuzzy searching, rather than the problem in the title, which is fine - that ended up being half the problem - and it was decided that a new issue would be made for the other half of the problem.

However, I hadn't bothered to make a new issue because my requirements had changed, and thus I don't need the functionality anymore. So @quagmt I'd recommend creating a new issue.

@epage perhaps I should change the title of this issue?

**Comment by weihanglo** [comment]
Feel free to open a new issue and change the title of this. Happy new year people!
